### ISSUE TITLE
Whatsapp update available

### ISSUE THREAD

**Issue by LiBOSS-CZ** [TheGoddessInari/hamsket#75]
### Steps to reproduce
1. Add Whatsapp account
2. Reload all services/Restart Rambox/Restart computer

### Expected behavior
Once click to update Whatsapp.

### Actual behavior
Nearly every restart of Rambox you have to update Whatsapp. (not everytime, but 75%)
I am using two Whatsapp account. 

![wweb](https://user-images.githubusercontent.com/41294900/50609188-429b3d80-0ecf-11e9-86d8-afd0c957f416.png)

I thinks it is starts to appeared after correcting the Issue #72

### ENVIRONMENT
> Rambox-OS: 0.5.18
> OS: Windows 10 Pro 1809
> Arch: x64
> BuildVersion: 0172bec

**Comment by TheGoddessInari** [comment]
I haven't been able to reproduce this,  WhatsApp Web says it's version 0.3.1846.
WhatsApp, why do you hate us so?

**Comment by LiBOSS-CZ** [comment]
Maybe it is because of two accounts logged in simultaneously. I have the same version as you. Before update it is same version as after update.

**Comment by wefalltomorrow** [comment]
I've been having that messages on and off for months, just click it and refresh to make it go away temporarily.

**Comment by M0UL** [comment]
> 
> 
> I've been having that messages on and off for months, just click it and refresh to make it go away temporarily.

same here

**Comment by LiBOSS-CZ** [comment]
> 
> 
> I've been having that messages on and off for months, just click it and refresh to make it go away temporarily.

Yes, I know, that you just click, but you do not want to do it almost every start of Rambox (for me twice)

**Comment by TheGoddessInari** [comment]
Don't suppose this is fixed on `master`?

If WhatsApp is freaking out on their end because someone's using multiple WhatsApp accounts, I'm not sure what can be done. None of the storage is shared, the tabs can't communicate with each other, so it's entirely on their end.

I don't seem to've asked before, but does this occur at all on other apps, on Android with features like dual messenger or sandboxing?

Maybe I'm overthinking it and there'd be some way to simply block the notice, since if you're on the web version, presumably there **isn't** an update unless you've been running a single session forever.

**Comment by LiBOSS-CZ** [comment]
On Android I do not have any problems with dual Whatsapp. Normally working as expected. As you said, it could be WhatsApp-side problem, but right now I have try Rambox CE and this issue does not occured.

**Comment by TheGoddessInari** [comment]
I've pushed a commit to `testing` that **tries** to see if we can remove the slowed timer blacklist from WhatsApp as per @JakobStruye's comment in #42.

This could possibly affect #31 too, so I'd like testing on WhatsApp to make sure it doesn't regress as well as to see if it fixes this.

**Comment by JakobStruye** [comment]
If you want to confirm quickly whether it _could_ be a fix for any WhatsApp issues, you could add something like
```
if (func_args.length > 0) {
    console.log("Now properly passing arguments in setTimeout");
}
```
to the override of setTimeout.

If that's never logged, there's no chance that fix will change anything for WhatsApp.

**Comment by TheGoddessInari** [comment]
Also a good thing to think of.
It logs it a number of times, on both login and when loading normally.
The source debugger's definitely working better than in electron 1.x, but I'm still somewhat mystified by it, so I can't reasonably say that it will or won't fix X, just that it's worth testing.

**Comment by LiBOSS-CZ** [comment]
I have tried build 204 - e311167 with same bad result:-(. I have noticed, that second whatsapp account is loading slower than the first one and it does not matter on tab order. Issues does not occur if I put Whatsapp accounts as 1st and 2nd tab. 1st an 3rd sometimes for the second account. With Rambox CE it is working as 1st and 5th tab. It seems to me that it has something to do with duration of loggin in.

**Comment by joe776** [comment]
Just FYI, happens for me with just a single Whatsapp account (I do have 3 other messengers however). Macbook Pro Mid'14 with lots of things running, so not the fastest machine.

**Comment by LiBOSS-CZ** [comment]
Now I am running WA and WA Bussiness account with same behaviour. I do not think that this is CPU power related. I am using it on i7 7700HQ with 16GB RAM and NVMe SSD.

**Comment by TheGoddessInari** [comment]
Since this appears to affect everyone but me, use the developer tools and see if you can come up with a simple JS or CSS to hide the update notification and I'll see if I can include it.

**Comment by joe776** [comment]
Just tried to have a look in your preview build, but that doesn't really work for me. I guess it clashes with the "old" Rambox version (there were also some exceptions in the dev console).
In the old version (pre-fork) i only get to the iframe which contains Whatsapp, but cannot access anything within it
So maybe this doesn't happen for you, Inari, because you do not use the "old" version of Rambox?

**Comment by TheGoddessInari** [comment]
I don't use the upstream version of Rambox, no.

If anyone wants this fixed for 0.5.18, which is ready-ish now except for this and #93,

Old data in the WhatsApp container will likely cause false-positives of problems, as stated many times. This is due to the design of WhatsApp. Other webapps tend to handle this situation themselves.

Containers/tabs can't talk with each other, so I'm not sure how two accounts could cause this, nor could I likely make two accounts myself. Opening the same account with two instances of WhatsApp Web for the same account definitely doesn't cause this for me.

WhatsApp has always been screwy with electron-based apps, and has been totally unresponsive to bug reports or requests of any kind in the last year.

**Comment by TheGoddessInari** [comment]
This should be resolved in 4aab913. If it's not, let me know and I can add a version override.

I'd prefer not to have to do that, of course, given that **all** of this is WhatsApp's fault for using bad User Agent string sniffing instead of feature detection.

**Comment by MatMaul** [comment]
It doesn't seem to be solved for me.

**Comment by TheGoddessInari** [comment]
As far as I can tell, it sometimes says this because it thinks that the version of Chrome in-use is too old, instead of letting Google handle Chrome updates. Since it's a redundant warning, and electron doesn't update as fast as WhatsApp likes, the best way to handle this would probably be to simply suppress that particular warning by default, but as it simply refuses(!) to pop up for me, I can't isolate the elements necessary to cleanly block it by default (without blocking anything else) using the CSS or JS as appropriate.

I used to block the useless "helper bar" at the bottom of Skype for the same reason because it simply wastes a fair amount of vertical space for no benefit, but they changed how things are arranged and there's nothing unique about it to block.

Lying about the version would require constant maintenance and they would possibly try to invoke features that the Electron version of Chromium simply doesn't have, since WhatsApp parses user agents instead of using proper feature detection, and totally ignores all reports sent to them, no matter the severity.

**Comment by TheGoddessInari** [comment]
The `testing` branch will currently be on Electron 5.0.0, which uses Chrome v73, which should be new enough for WhatsApp.

Let me know if you find any problems with electron 5 because it'll probably become the default if there aren't more wrinkles to iron out.

**Comment by MatMaul** [comment]
Confirmed, no more update nagging on `testing` branch :)

**Comment by LiBOSS-CZ** [comment]
I have tried on latest master (d42dc5c) also with Chrome v73, but still appears. It is appearing a lot less, but still appears (1 in 10 maybe).

**Comment by LiBOSS-CZ** [comment]
This problem is back:-( Even with the lates Hamsket version 9cc1a71

**Comment by TheGoddessInari** [comment]
This is surreal. Chrome 76 is the latest release version (and as usual, I can't replicate).

I think the problem is...we have no idea what WhatsApp is actually demanding here. They're totally opaque about it, and refuse to answer any support tickets.

**Comment by LiBOSS-CZ** [comment]
I think, that problems is in two Whatsapp (I am using Whatsapp and Whatsapp Bussiness) account signing in simultaneously. Is any possibility how to delay singing in for one account?

**Comment by YerayAlonso** [comment]
Add the following code in the "advanced functions" box at the whatsapp settings to hide that alert:

```javascript
(()=>{ let css=`
._3O0po {
    display: none;
}
`;

var ss = document.createElement("style");
ss.type = "text/css"; ss.innerHTML = css;
document.getElementsByTagName("head")[0].appendChild(ss); })();
```

**Comment by TheGoddessInari** [comment]
@YerayAlonso Mind if I add that to official?

**Comment by francwalter** [comment]
I have it too, sometimes more often, sometimes not at all for a while, at the moment with my Windows 10 tablet and latest Hamsket nightly (though I think it is unrelated with Hamsket).
When I tap the update button, I get that silly Update Google Chrome screen, so it doesnt help at all.
I will live with this notification and will ignore it. 
Maybe I can get rid of it with some CSS in the settings 😏

**Comment by francwalter** [comment]
I think it is related to that Beta, quite new now, must have some totally different CSS-Classes. That "_3O0po" Class is not in use obviously, that trick from YerayAlonso doesnt work anymore.
I tried the ID "side" which I got from the "real" WA-Web-Version, but it is not known in Hamsket's WA. I thought with "side" I could slide to the bad ID with next-sibling or something like that, but if it is unknown I am over with my latin.
I tried in DevTools things like:

`for(i=1;i++;i==400) document.getElementsByTagName("DIV")[i].style.color = "red";`

But this affects only all the Hamsket Window around the WA-content. No way to get to that.

If I only knew how to find the actual class or ID of that damned element with that damned "Update available".
I guess there are several developers employed at Facebook just to make it harder to use WA in any other environment than they want to.

**Comment by DeLuke84** [comment]
I can remove the element from dev tools with:

const elements = document.getElementsByClassName("_3z9_h");
elements[0].remove();

this is removing the span element where the child element is displaying the "Update available"

Unfortunately, when I pass this code to the advanced section, it doesn't work anymore... 

any ideas why?

**Comment by DeLuke84** [comment]
With that piece of code, the span element which contains the 'Update available' message can be removed.

(() => {
  setTimeout(() => document.getElementsByClassName("_3z9_h")[0].remove(), 7000);
})();

A better solution would be to have some kind of event handler instead of a timeout, which check's whether the page has been loaded. 

Since I'm not a JS expert, I don't know how to do it.

**Comment by francwalter** [comment]
The class
`_3z9_h`
might be a randomly given classname, on my system it is unknown :(
If I do in the console of the devtools some:
`document.getElementsByClassName("_3z9_h")[0]`
I can see that it is undefined.

**Comment by francwalter** [comment]
I had this issue again on Ubuntu and Hamsket 0.6.5 (6e2..., last nightly ever from 6. Aug. 2022) and still it is not possible to get any ID or class of this element with "Update available", because it is in the webview element, where I cannot get into with the Dev. Tools.
I would need this message in a real browser but there it never appears.

So my way to get temporarily rid of the message, which takes the most important space (the top of my contact's chats) is still to refresh, often several times. Then it is gone for a while.

I still hope that one day anybody could tell me, how to scrutinize that fu...ng element with Web dev. Tools, get into that Webview element in Hamsket, but I think it is just not possible.

**Comment by francwalter** [comment]
> ...
> I still hope that one day anybody could tell me, how to scrutinize that fu...ng element with Web dev. Tools, get into that Webview element in Hamsket, but I think it is just not possible.

At least I found a way in Ferdium, see my issue and own answer here: [Update available - get CSS or TagName of DIV](https://github.com/vednoc/dark-whatsapp/issues/199)

If my code with the chat-butterbar works a while:
`document.querySelectorAll('[data-testid="chat-butterbar"]')[0].style.display = "none";`
I will post it here completely...

**Comment by jschiefner** [comment]
@francwalter i found out again how to open the dev tools for a service, not for the whole hamsket interface, right-click on the `WhatsApp` Service in the menu bar and hit `Toggle Developer Tools` from there.

With this, the message can be hidden with one of the following ways:

```javascript
document.querySelector('span[data-testid="alert-update"]').parentElement.parentElement.parentElement.parentElement.style.display = "none"
// or
document.querySelector('span[data-testid="alert-update"]').parentElement.parentElement.parentElement.style.display = "none"

// or as mentioned before
document.querySelector('span[data-testid="chat-butterbar"]').style.display = "none"
```

To reliably hide the message i inject the following javascript:

```javascript
const interval = setInterval(hideUpdateMessage, 1000);

function hideUpdateMessage() {
  const updateMessage = document.querySelector('span[data-testid="chat-butterbar"]');
  if (updateMessage !== undefined) {
    updateMessage.style.display = "none";
    clearInterval(interval);
  };
};
```

**Comment by francwalter** [comment]
> If my code with the chat-butterbar works a while: `document.querySelectorAll('[data-testid="chat-butterbar"]')[0].style.display = "none";` I will post it here completely...

Yes, this works since a while, got rid of that annoying update butter bar.
I even put it to some computer where I still use Hamsket, put the code without the ferdium wrapper to the JS insert setting window. It works!
